{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import pathlib\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import glob\n",
    "from Bio.Seq import Seq\n",
    "from Bio import SeqIO\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "from Bio.SeqIO.FastaIO import SimpleFastaParser\n",
    "import subprocess \n",
    "from BCBio import GFF\n",
    "import re \n",
    "import matplotlib.pyplot as plt "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Fasta file writer \n",
    "def FastaFile(gene_id,seq,outname):\n",
    "    initial_path = os.getcwd()\n",
    "    with open(outname,'a') as f:\n",
    "        f.write(\">\"+gene_id+\"\\n\"+seq+\"\\n\") \n",
    "    os.chdir(initial_path)\n",
    "\n",
    "\n",
    "def extract_upstream_sequence(contig_id, start, end,genome_dict, upstream_length=3000):\n",
    "    \"\"\"\n",
    "    Extracts the sequence upstream of a gene given its genomic location.\n",
    "    contig_id: str, ID of the contig containing the gene.\n",
    "    start: int, 1-based start position of the gene.\n",
    "    end: int, 1-based end position of the gene.\n",
    "    upstream_length: int, length of sequence to extract upstream of the gene.\n",
    "    Returns: str, upstream sequence of the gene.\n",
    "    \"\"\"\n",
    "\n",
    "    # Get sequence object for contig\n",
    "    contig_seq = genome_dict[contig_id].seq\n",
    "    # Compute start and end positions of upstream sequence\n",
    "    upstream_start = max(start - upstream_length, 0)\n",
    "    upstream_end = start - 1\n",
    "    # Extract upstream sequence\n",
    "    upstream_seq = contig_seq[upstream_start:upstream_end]\n",
    "    return str(upstream_seq)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Class Class()\n",
    "class Class():\n",
    "    def load(self,filename): #load proteome data (single or multiple fasta files) \n",
    "        # Create an empty DataFrame with the desired column names\n",
    "        self.input_seq_db  = pd.DataFrame(columns=[\"gene_id\", \"seq\", \"len\"])\n",
    "        # Open the FASTA file and iterate over its records\n",
    "        with open(filename) as in_handle:\n",
    "            for title, seq in SimpleFastaParser(in_handle):\n",
    "                # Add the current record to the DataFrame\n",
    "                len_seq = len(seq)\n",
    "                gene_id = title.split()[0]\n",
    "                self.input_seq_db = pd.concat([self.input_seq_db, pd.DataFrame([[gene_id, seq, len_seq]], columns=[\"gene_id\", \"seq\", \"len\"])])\n",
    "        self.input_seq_db.set_index(keys=self.input_seq_db[\"gene_id\"],inplace=True)\n",
    "        return self \n",
    "    def combine_df(self,filename): #load proteome data (single or multiple fasta files) \n",
    "        with open(filename) as in_handle:\n",
    "            for title, seq in SimpleFastaParser(in_handle):\n",
    "                # Add the current record to the DataFrame\n",
    "                len_seq = len(seq)\n",
    "                gene_id = title.split()[0]\n",
    "                self.input_seq_db = pd.concat([self.input_seq_db, pd.DataFrame([[gene_id, seq, len_seq]], columns=[\"gene_id\", \"seq\", \"len\"])])\n",
    "        return self \n",
    "\n",
    "    def hmmer_cyp(self,input_fasta,):\n",
    "        # Run HMMER\n",
    "        subprocess.check_output(['hmmsearch', \"-E\",\"0.1\", \"--tblout\", \"hmm_hits.txt\",\"Pfam-A.hmm\",input_fasta])\n",
    "        # Process the HMMER output and add hits to the DataFrame\n",
    "        with open(\"hmm_hits.txt\") as f:\n",
    "            for line in f:\n",
    "                if not line.startswith('#'):\n",
    "                    fields = line.strip().split()\n",
    "                    e_value = fields[4]\n",
    "                    gene_id = fields[0]\n",
    "                    domain_name = fields[3]\n",
    "                    if domain_name in ['PF00067.25', 'PF00067']:\n",
    "                        mask = self.input_seq_db[\"gene_id\"] == gene_id\n",
    "                        if mask.any():\n",
    "                            self.input_seq_db.loc[mask, \"Hmmer_CYP\"] = e_value\n",
    "        # Remove the HMMER output file\n",
    "        os.remove(\"hmm_hits.txt\")\n",
    "        return self \n",
    "\n",
    " \n",
    "    def blastp_comparison(self, query, subject):\n",
    "        # Run blastp\n",
    "        subprocess.check_output(['blastp', \"-query\", query,\"-subject\",subject, \"-out\",\"blastp_output.txt\",\"-outfmt\",\"6\",\"-max_target_seqs\",\"1\",\"-evalue\",\"0.1\"])\n",
    "        # Process the blastp output and add hits to the DataFrame\n",
    "        with open('blastp_output.txt') as f:\n",
    "            for line in f:\n",
    "                fields = line.split()\n",
    "                gene_id = fields[0]\n",
    "                subject_id = fields[1]\n",
    "                percentage = fields[2]\n",
    "                e_value = fields[10]\n",
    "                mask = self.input_seq_db[\"gene_id\"] == gene_id\n",
    "                header = subject_id+\"aa identity\"\n",
    "                if mask.any():\n",
    "                    self.input_seq_db.loc[mask, header] = str(percentage)\n",
    "                    self.input_seq_db.loc[mask, \"evalue\"] = str(e_value)\n",
    "        os.remove(\"blastp_output.txt\")\n",
    "        return self\n",
    "    \n",
    "    def blastp_label(self,query,subject):\n",
    "        # Run blastp of the transcriptome object against a target protein, label the e-value and %homology of hits \n",
    "        subprocess.check_output(['blastp', \"-query\", query,\"-subject\",subject, \"-out\",\"blastp_output.txt\",\"-outfmt\",\"6\",\"-max_target_seqs\",\"1\",\"-evalue\",\"0.1\"])\n",
    "        # Process the blastp output and add hits to the DataFrame\n",
    "        with open('blastp_output.txt') as f:\n",
    "            for line in f:\n",
    "                fields = line.split()\n",
    "                gene_id = fields[0]\n",
    "                subject_id = fields[1]\n",
    "                percentage = fields[2]\n",
    "                e_value = fields[10]\n",
    "                mask = self.input_seq_db[\"gene_id\"] == gene_id\n",
    "                if mask.any():\n",
    "                    self.input_seq_db.loc[mask, \"CYP blastp - CYP Name\"] = subject_id\n",
    "                    self.input_seq_db.loc[mask, \"CYP blastp - aa perc\"] = percentage\n",
    "                    self.input_seq_db.loc[mask, \"CYP blastp - e value\"] = e_value\n",
    "        os.remove(\"blastp_output.txt\")\n",
    "        return self  \n",
    "    \n",
    "    def get_distribution_identity(self,header_name):\n",
    "        #Use this after using blastp_label to plot the results \n",
    "        plt.hist(self.input_seq_db[header_name].str.split(\",\").str[0].astype(float), bins=20, range=(0, 100))\n",
    "        plt.xlabel(\"Percentage Amino Acid Identity\")\n",
    "        plt.ylabel(\"Count\")\n",
    "        plt.title(\"Distribution of Percentage Amino Acid Identity\")\n",
    "        plt.show()\n",
    "        return plt "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "tacca = Class() \n",
    "tacca.load(\"tchantrieri_proteome.fasta\")\n",
    "\n",
    "with open(\"hmm_hits_tps.txt\") as f:\n",
    "    for line in f:\n",
    "        if not line.startswith('#'):\n",
    "            fields = line.strip().split()\n",
    "            e_value = fields[4]\n",
    "            gene_id = fields[0]\n",
    "            domain_name = fields[3]\n",
    "            if domain_name in ['PF00432.24']:\n",
    "                mask = tacca.input_seq_db[\"gene_id\"] == gene_id\n",
    "                if mask.any():\n",
    "                    tacca.input_seq_db.loc[mask, \"Hmmer_prenyltr\"] = e_value"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import requests\n",
    "from bs4 import BeautifulSoup\n",
    "\n",
    "# Define a function that checks whether there are SRA experiments available for a given taxon\n",
    "def check_sra(txid, seq_type):\n",
    "    url = f\"https://www.ncbi.nlm.nih.gov/sra?term={txid}%20{seq_type}\"\n",
    "    response = requests.get(url)\n",
    "    soup = BeautifulSoup(response.content, \"html.parser\")\n",
    "    title = soup.find(\"title\").text\n",
    "    print(title)\n",
    "    if \"No items found\" in title:\n",
    "        return \"No\"\n",
    "    else: \n",
    "        print('hi')\n",
    "        return \"Yes\"\n",
    "\n",
    "\n",
    "# Read in the CSV file and specify \"Plant Taxonomy\" as the column label\n",
    "df = pd.read_csv('unique_plant_species.csv')\n",
    "print(df)\n",
    "\n",
    "# Add a new column for DNA-seq experiments and populate it with \"Yes\" or \"No\"\n",
    "df['DNA-seq'] = df['Plant Species'].apply(lambda x: check_sra(x, \"DNA\"))\n",
    "\n",
    "# Add a new column for RNA-seq experiments and populate it with \"Yes\" or \"No\"\n",
    "df['RNA-seq'] = df['Plant Species'].apply(lambda x: check_sra(x, \"RNA\"))\n",
    "\n",
    "# Save the updated DataFrame to a new CSV file\n",
    "df.to_csv('sra_experiments.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#BLASTING AGAINST NELSON MOSS DB \n",
    "path_input = \"/Users/bustin/Desktop/capstone/MOSSBIOINFO/transcriptomics/march25/biojustin/data/datadep\"\n",
    "os.chdir(path_input)\n",
    "nelson = BioBustin()\n",
    "nelson.load(filename=\"cyp_filtered_6000aa_wref.fasta\",path_input=path_input)\n",
    "for index,row in nelson.input_seq_db.iterrows():\n",
    "    seq = row[\"seq\"]\n",
    "    query = row[\"gene_id\"]\n",
    "    outname = subject+\".fasta\"\n",
    "    FastaFile(gene_id=query,seq=seq,outname= outname)\n",
    "    nelson.blastp_label(query=outname, subject=\"cyp_db_full.fasta\")\n",
    "    os.remove(outname)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "nelson.input_seq_db.loc[:,\"CYP blastp - CYP Name\":\"CYP blastp - e value]\"].to_csv(\"nelson_label_6000.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "path_input = \"/Users/bustin/Desktop/capstone/MOSSBIOINFO/transcriptomics/march25/biojustin/data/datadep\"\n",
    "\n",
    "test1 = BioBustin()\n",
    "test1.load(filename=\"cyp_filtered_6000aa_wref.fasta\",path_input=path_input)\n",
    "test1.hmmer_cyp(pfam=\"PF00067.hmm\",input_fasta=\"cyp_filtered_6000aa_wref.fasta\",path_input=path_input)\n",
    "test1.blastp_comparison(query=\"cyp_filtered_6000aa_wref.fasta\",subject=\"CpCYP964A1_p.fasta\")\n",
    "test1.EC_CLEAN(input_file=\"cyp_filtered_6000aa_wref.fasta\")\n",
    "print(test1.input_seq_db)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Make the full nelson cyp database fasta files \n",
    "with open(\"cyp_db_full.fasta\",'a') as f:\n",
    "    for i in [\"Chlorophyte_algae\",\"Charophyte_algae\",\"MossRaw\",\"Liverworts\",\"Hornworts\",\"Lycophytes\",\"Fern\",\"Gymnosperm\"]:\n",
    "        df = pd.read_excel(\"nelson_cyp_db.xlsx\", i)\n",
    "        for index, row in df.iterrows():\n",
    "            seq = row[\"Sequence\"]\n",
    "            gene_id = str(row[\"CYP name\"]) + \",\" + str(row[\"Species\"]) + \",\" + i\n",
    "            f.write(\">\"+gene_id+\"\\n\"+seq+\"\\n\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Make just moss db \n",
    "xls = pd.ExcelFile('NELSON_DB.xlsx')\n",
    "df = pd.read_excel(xls, \"MossRaw\")\n",
    "with open(\"cyp_db_moss.fasta\",'w') as f:\n",
    "    for index, row in df.iterrows():\n",
    "        seq = row[\"Sequence\"]\n",
    "        gene_id = row[\"CYP name\"] + \",\" + row[\"Species\"] + \",\" + \"MossRaw\"\n",
    "        f.write(\">\"+gene_id+\"\\n\"+seq+\"\\n\")\n",
    "        \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#OS STUFF PATH STUFF \n",
    "#print(os.getcwd())\n",
    "#fasta_file = \"CpCYP964A1_p.fasta\"\n",
    "#print(os.path.isfile(fasta_file))\n",
    "#st = os.stat(fasta_file)\n",
    "#print(st)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": ".venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.6"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
